### PR TITLE
units: add a tpm2.target synchronization point and small generator that pulls in

### PR DESCRIPTION
**Pull request by poettering**
Material for v256

Fixes: #30164

### CODE REVIEW

File: test/units/testsuite-64.sh [hunk 1]
```diff
@@ -148,7 +148,7 @@ check_device_units() {(
         if ! check_device_unit "$log_level" "$path"; then
            return 1
         fi
-    done < <(systemctl list-units --all --type=device --no-legend dev-* | awk '$1 !~ /dev-tty.+/ { print $1 }' | sed -e 's/\.device$//')
+    done < <(systemctl list-units --all --type=device --no-legend dev-* | grep plugged | awk '$1 !~ /dev-tty.+/ { print $1 }' | sed -e 's/\.device$//')
```
Comment (dtardon) [inline test/units/testsuite-64.sh]: Ugh, `awk`, `grep` and `sed` in a single pipeline... Please just add an additional condition to the `awk` pattern, i.e., `awk '$1 !~ /dev-tty.+/ && $4 == "plugged" { print $1 }'`

(While at it, the `sed` call could be dropped too by replacing `print $1` by `sub(/.device$/, "", $1); print $1` .)

Comment (poettering) [reply test/units/testsuite-64.sh]: you are forgetting though that i can read the combination of sed, awk and grep, but not the complex awk expression you suggested ;-)

Comment (dtardon) [reply test/units/testsuite-64.sh]: At least the first suggestion should be understandable, shouldn't it? I mean, if people are able to guess that `$1 !~ /dev-tty.+/` means regex match against the first column of the input, then a match against two columns at once shouldn't be that hard to figure out either.

**Comment by poettering** [comment]
Addressed @dtardon's point.

Hey, @bluca any chance you can give this a blessing?

File: src/tpm2-setup/tpm2-generator.c [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                        return 0;
+                }
+
+                if (!FLAGS_SET(support, TPM2_SUPPORT_SYSTEM|TPM2_SUPPORT_SUBSYSTEM|TPM2_SUPPORT_LIBRARIES)) {
+                        log_debug("Not generating tpm2.target synchronization point, as userspace support for TPM2 is not complete.");
+                        return 0;
+                }
+        }
+
+        r = generator_add_symlink(arg_dest, SPECIAL_SYSINIT_TARGET, "wants", SYSTEM_DATA_UNIT_DIR "/" SPECIAL_TPM2_TARGET);
```
Comment (bluca) [inline src/tpm2-setup/tpm2-generator.c:61]: shouldn't this be skipped in `arg_tpm2_wait == 0`?

Comment (poettering) [reply src/tpm2-setup/tpm2-generator.c:61]: uh, yes, absolutely

**Comment by poettering** [comment]
Force pushed a new version with the missing check added.

**Review by bluca — APPROVED** [review]

File: src/tpm2-setup/tpm2-generator.c [hunk 3]
```diff
@@ -0,0 +1,80 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "generator.h"
+#include "proc-cmdline.h"
+#include "special.h"
+#include "tpm2-util.h"
+#include "parse-util.h"
```
Comment (YHNdnzj) [inline src/tpm2-setup/tpm2-generator.c:7]: Not sorted?

**Pull request merged**